cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkCredentialsUpdater.java

```diff
@@ -0,0 +1,239 @@
+
+  public SparkCredentialsUpdater(Supplier<Credentials> credentialsSupplier,
+                                 Location credentialsDir, String fileNamePrefix, long updateIntervalMs,
+                                 long cleanupExpireMs, int minFilesToKeep) {
+    Preconditions.checkArgument(updateIntervalMs > 0, "Update interval must be positive");
+    Preconditions.checkArgument(cleanupExpireMs > 0, "Cleanup expire time must be positive");
+    Preconditions.checkArgument(minFilesToKeep >= 0, "Minimum files to keep must be non-zero");
+
+    this.credentialsSupplier = credentialsSupplier;
+    this.credentialsDir = credentialsDir;
+    this.fileNamePrefix = fileNamePrefix;
+    this.updateIntervalMs = updateIntervalMs;
+    this.cleanupExpireMs = cleanupExpireMs;
+    this.minFilesToKeep = minFilesToKeep;
+    this.fileNamePattern = Pattern.compile(fileNamePrefix + SPARK_YARN_CREDS_COUNTER_DELIM + "(\\d+)");
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    scheduler = Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("credentials-updater"));
+    scheduler.submit(this).get();
+
+    LOG.info("Credentials updater started");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    ExecutorService executor = scheduler;
+    if (executor != null) {
+      executor.shutdownNow();
+    }
+
+    LOG.info("Credentials updater stopped");
+  }
+
+  @Override
+  public void run() {
+    long nextUpdateTime = updateIntervalMs;
+    try {
+      if (generation == 0) {
+        generation = findLatestGeneration();
+      }
+
+      // Write to the next generation file. It's ok to skip some generation if the write failed.
+      generation++;
+
+      Location credentialsFile = credentialsDir.append(fileNamePrefix + SPARK_YARN_CREDS_COUNTER_DELIM + generation);
+      Location tempFile = credentialsDir.append(credentialsFile.getName() + SPARK_YARN_CREDS_TEMP_EXTENSION);
+
+      // Writes the credentials to temp location, then rename to the final one
+      Credentials credentials = credentialsSupplier.get();
+      try (DataOutputStream os = new DataOutputStream(tempFile.getOutputStream("600"))) {
+        credentials.writeTokenStorageToStream(os);
+      }
+
+      if (!credentialsFile.equals(tempFile.renameTo(credentialsFile))) {
+        throw new IOException("Failed to rename from " + tempFile + " to " + credentialsFile);
+      }
+
+      LOG.info("Credentials written to {}", credentialsFile);
+
+      // Schedule the next update.
+      // Use the same logic as the Spark executor to calculate the update time.
+      nextUpdateTime = getNextUpdateDelay(credentials);
+
+      LOG.debug("Next credentials refresh at {}ms later", nextUpdateTime);
+      scheduler.schedule(this, nextUpdateTime, TimeUnit.MILLISECONDS);
+      cleanup();
+
+    } catch (Exception e) {
+      // Retry time is the min(1 minute, update interval)
+      long retryDelay = Math.min(60000, nextUpdateTime);
```

Comment:
It is in case if there is any failure of getting the credentials. The retry shouldn't be too short, as typically credentials expire in hours to be actually useful. So retrying in a minute seems like a logical arbitrary choice without the need of introducing yet another configuration which rarely needed.